### PR TITLE
BatchLoader fails with expired cache entry

### LINKED ISSUE #2269: BatchLoader fails with expired cache entry
**Describe the bug**

Each batch loader creates many cache entries related to that model. If one of these cache entries expires, the corresponding key will be added to the batch loader after the batch loader resolve function. Therefore there are some keys inside the batch loader without results. [This line](https://github.com/nuwave/lighthouse/blob/df53035320bf8945b2d02ccc24ac8fde0e3146a8/src/Execution/BatchLoader/RelationBatchLoader.php#L62) throws an error.

**Expected behavior/Solution**

The batch loader resolve function must wait until all the model keys are loaded. 

**Steps to reproduce**

Fail test added:
testBatchLoaderFailsWithExpiredCacheEntry();

**Output/Logs**

<details><summary>Click to expand</summary>

```
# Add in log output/error messages here
```

</details></br>

**Lighthouse Version**
latest

### PR DESCRIPTION
**Pull request by mostafa-rz**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

Resolves #2269 

**Changes**

- add failed test

**Breaking changes**

### CODE REVIEW

File: tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php [hunk 1]
```diff
@@ -570,4 +572,93 @@ public function testCombineEagerLoadsThatAreTheSameRecursively(): void
...
+                        user {
+                            id
+                        }
+                    }
+                }
+            }
+            ');
+        } catch (\Exception $exception) {
+            $this->fail();
+        }
```
Comment (spawnia) [inline tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:642]: Just let this throw

File: src/Execution/BatchLoader/RelationBatchLoader.php [hunk 2]
```diff
@@ -53,6 +53,7 @@ public function load(Model $model): Deferred
     {
         $modelKey = ModelKey::build($model);
         $this->parents[$modelKey] = $model;
+        $this->hasResolved = false;
```
Comment (spawnia) [inline src/Execution/BatchLoader/RelationBatchLoader.php:56]: I am confused why this is necessary, can you explain?

Comment (mostafa-rz) [reply src/Execution/BatchLoader/RelationBatchLoader.php:56]: The first request without retrieving anything from the cache execution is like this:

1. posts
2. creating posts BatchLoader
3. resolving posts BatchLoader
4. comments 
5. creating comments BatchLoader 
6. resolving comments BatchLoader

The second request with retrieving from the cache:

1. posts
2. creating comments BatchLoader
3. creating posts BatchLoader
4. resolving comments BatchLoader
5. resolving posts BatchLoader
6. comments 
7. adding new keys to the current resolved comments BatchLoader

I know that this is not the RelationBatchLoader problem, but by adding this line I force the batch loader to be run again if there are new keys inside the batch loader.

Comment (mostafa-rz) [reply src/Execution/BatchLoader/RelationBatchLoader.php:56]: I don't have any idea why a comment batch loader must be created when we are in the step to resolve the posts.

File: tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php [hunk 3]
```diff
@@ -570,4 +572,86 @@ public function testCombineEagerLoadsThatAreTheSameRecursively(): void
         // @phpstan-ignore-next-line Of course this terminates...
         $this->assertSame(3, $queryCount);
     }
+
+    public function testBatchLoaderFailsWithExpiredCacheEntry(): void
+    {
+        $this->schema /** @lang GraphQL */
+            = '
```
Comment (spawnia) [inline tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:579]:
```suggestion
        $this->schema = /** @lang GraphQL */ '
```

File: tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php [hunk 4]
```diff
@@ -570,4 +572,86 @@ public function testCombineEagerLoadsThatAreTheSameRecursively(): void
...
+        $comments2 = factory(Comment::class, 3)->make();
+        foreach ($comments2 as $comment) {
+            assert($comment instanceof Comment);
+            $comment->user()->associate($user2);
+            $comment->post()->associate($post2);
+            $comment->save();
+        }
+
+        $firstRequest = $this
+            ->graphQL(/** @lang GraphQL */ '
```
Comment (spawnia) [inline tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:628]:
```suggestion
        $firstRequest = $this->graphQL(/** @lang GraphQL */ '
```

File: tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php [hunk 5]
```diff
@@ -570,4 +572,86 @@ public function testCombineEagerLoadsThatAreTheSameRecursively(): void
...
+                    }
+                }
+            }
+        }
+        ');
+
+        Cache::forget('lighthouse:Post:5:comments');
+
+        $secondRequest = $this
+            ->graphQL(/** @lang GraphQL */ '
```
Comment (spawnia) [inline tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:643]:
```suggestion
        $secondRequest = $this->graphQL(/** @lang GraphQL */ '
```

File: src/Execution/BatchLoader/RelationBatchLoader.php (hunk 2 above)
Comment (spawnia) [reply src/Execution/BatchLoader/RelationBatchLoader.php:56]: I think this fix leads to suboptimal duplicate queries - here is the output of the debugging statements I added:

```
"First request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:630
"select * from `posts` where `posts`.`deleted_at` is null" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `comments` where `comments`.`post_id` in (1, 2, 3, 4, 5, 6, 7, 8)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"Second request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:651
"select * from `users` where `users`.`id` in (1)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `comments` where `comments`.`post_id` in (5)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
```

Comment (mostafa-rz) [reply src/Execution/BatchLoader/RelationBatchLoader.php:56]: Do you have any idea why when cache directive is used, the order of field resolving is changed?

Comment (spawnia) [reply src/Execution/BatchLoader/RelationBatchLoader.php]: I figured out it was due to the relation batch loader returning a `Deferred` result, whereas retrieving a value from cache returns a result immediately. To smooth things over, I tried returning a `Deferred` from `CacheDirective` and found that it fixes the duplicate request for `users` and makes this fix unnecessary.

```
"First request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:630
"select * from `posts` where `posts`.`deleted_at` is null" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `comments` where `comments`.`post_id` in (1, 2, 3, 4, 5, 6, 7, 8)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"Second request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:651
"select * from `comments` where `comments`.`post_id` in (5)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
```

**Pull request merged**